File: Data Structures & Algorithms/max-area-of-island/submission-2.cpp
class Solution {
public:
        int dfs(int r, int c, vector<vector<int>>& grid) {
            if(r < 0 || c < 0 || r >= grid.size() || c >= grid[r].size()) {
                return 0;
            }
            if(grid[r][c] == 0) {
                return 0;
            }
            grid[r][c] = 0; //mark as visited
            return 1 + 
                dfs(r + 1, c, grid) + 
                dfs(r - 1, c, grid) +
                dfs(r, c + 1, grid) + 
                dfs(r, c - 1, grid);
        }

    int maxAreaOfIsland(vector<vector<int>>& grid) {

        int m = 0;
        for(int i = 0; i < grid.size(); i++) {
            for(int j = 0; j < grid[i].size(); j++) {
                int current = dfs(i, j, grid);
                if(current > m) {
                    m = current;
                }
            }
        }
        return m;

    }
};

File: Data Structures & Algorithms/word-ladder/submission-2.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        if wordList == []:
            return 0
        visited = set()
        graph = defaultdict(list)

        #Create the graph
        for i in range(len(wordList)):
            word = wordList[i]
            for j in range(len(word)):
                parent = word[:j] + '*' + word[j + 1:]
                graph[parent].append(word)
                graph[word].append(parent)
                if parent == beginWord[:j] + '*' + beginWord[j + 1:]:
                    graph[parent].append(beginWord)
                    graph[beginWord].append(parent)
                

        #Run BFS from the starting word
        queue = deque([(beginWord, 1)])
        visited.add(beginWord)
        while queue:
            current, length = queue.popleft()
            if current == endWord:
                return length

            for i in range(len(current)):
                pattern = current[:i] + '*' + current[i + 1:]

                for neighbor in graph[pattern]:
                    if neighbor not in visited:
                        queue.append((neighbor, length + 1))
                        visited.add(neighbor)

        return 0


